### ISSUE TITLE
Training with custom dataset fails.

### ISSUE THREAD

**Issue by DeltaTesting** [zzh8829/yolov3-tf2#63]
Hi,

When i try to train the model with a custom dataset, I get the following error:

`W tensorflow/core/framework/op_kernel.cc:1546] OP_REQUIRES failed at sparse_xent_op.cc:90 : Invalid argument: Received a label value of -1 which is outside the valid range of [0, 80)`

I'm not sure if i am creating my tfrecord wrong or if the error comes from anywhere else.
I create my tf records with the following script:

`
def get_example(line):
    class_text = []
    labels = []
    xmin = []
    ymin = []
    xmax = []
    ymax = []
    truncated = []
    poses = []
    difficult_obj = []

    line = line.split('"')

    image_path = line[-2]
    with tf.io.gfile.GFile(image_path, 'rb') as fib:
      image_encoded = fib.read()
    w,h = (tf.keras.preprocessing.image.load_img(image_path)).size
    key = hashlib.sha256(image_encoded).hexdigest()
    
    line = line[-1].split(' ')

    for item in line[1:]:
        item = item.split(',')
        xmin.append(float(item[0]))
        ymin.append(float(item[1]))
        xmax.append(float(item[2]))
        ymax.append(float(item[3]))

        class_text.append(item[4].encode('utf8'))
        labels.append(int(item[4]))
        print(int(item[4]))

        truncated.append(0)
        difficult_obj.append(0)
        poses.append("Unspecified".encode('utf8'))

    example = tf.train.Example(features=tf.train.Features(feature={
      'image/height': int64_feature(h),
      'image/width': int64_feature(w),
      'image/filename': bytes_feature(image_path.encode('utf8')),
      'image/source_id': bytes_feature(image_path.encode('utf8')),
      'image/key/sha256': bytes_feature(key.encode('utf8')),
      'image/encoded': bytes_feature(image_encoded),
      'image/format': bytes_feature('jpeg'.encode('utf8')),
      'image/object/bbox/xmin': float_list_feature(xmin),
      'image/object/bbox/xmax': float_list_feature(xmax),
      'image/object/bbox/ymin': float_list_feature(ymin),
      'image/object/bbox/ymax': float_list_feature(ymax),
      'image/object/class/text': bytes_list_feature(class_text),
      'image/object/class/label': int64_list_feature(labels),
      'image/object/difficult': int64_list_feature(difficult_obj),
      'image/object/truncated': int64_list_feature(truncated),
      'image/object/view': bytes_list_feature(poses),
    }))

    return example

train_writer = tf.io.TFRecordWriter('./data/train.tfrecord')
val_writer = tf.io.TFRecordWriter('./data/val.tfrecord')
with open(train_txt, 'r') as f:
    lines = f.read().split('\n')
    random.shuffle(lines)
    nLines = len(lines)
    nTrain = int(nLines * 0.8)
    print("creating train.tfrecord")
    for line in lines[:nTrain]:
        if len(line) > 0:
            example = get_example(line)
            train_writer.write(example.SerializeToString())

    print("creating val.tfrecord")
    for line in lines[nTrain:]:
        if len(line) > 0:
            example = get_example(line)
            val_writer.write(example.SerializeToString())
train_writer.close()
val_writer.close()
print("finish!")
`

**Comment by htzmedia** [comment]
Is your Dataset in PascalVOC Format (Image + XML-File)?

**Comment by DeltaTesting** [comment]
Not exactly, I have plain txt file instead of XML. But when creating the tfrecord this shouldn't be a problem, isn't it?

My .txt file has the following structure:
"image/path" xmin,ymin,xmax,ymax,objectid xmin,ymin,xmax,ymax,objectid ....

The x's and y's are in percentage (interval [0,1)).

**Comment by htzmedia** [comment]
I could post my code. But I am using xml-files. 
In the first step, I create a big csv-File with all Images and then I am converting it to tfrecord.

I am using a config.py for all my paths so you have to modify a bit. But for me everything works fine.
In the beginning I also had trouble with my tfrecord files.

```
import io
from PIL import Image
import hashlib
import tensorflow as tf
import os
import pandas as pd
import xml.etree.ElementTree as ET
import sys
from random import shuffle
from collections import namedtuple, OrderedDict
import config

def xml_to_csv(p):
    xml_list = []
    file_list = os.listdir(p)
    for filename in file_list:
        if filename.lower().endswith('.xml'):
            tree = ET.parse(p + '/' + filename)
            root = tree.getroot()
            for member in root.findall('object'):
                value = (root.find('filename').text, int(root.find('size')[0].text), int(root.find('size')[1].text),
                         member[0].text, int(float(member[4][0].text)), int(float(member[4][1].text)),
                         int(float(member[4][2].text)), int(float(member[4][3].text)))
                xml_list.append(value)
        column_name = ['filename', 'width', 'height', 'class', 'xmin', 'ymin', 'xmax', 'ymax']
        xml_df = pd.DataFrame(xml_list, columns=column_name)
    return xml_df


def split(df, group):
    data = namedtuple('data', ['filename', 'object'])
    gb = df.groupby(group)
    return [data(filename, gb.get_group(x)) for filename, x in zip(gb.groups.keys(), gb.groups)]

def bytes_feature(value):
  return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))

def bytes_list_feature(value):
  return tf.train.Feature(bytes_list=tf.train.BytesList(value=value))

def float_list_feature(value):
  return tf.train.Feature(float_list=tf.train.FloatList(value=value))

def int64_feature(value):
  return tf.train.Feature(int64_list=tf.train.Int64List(value=[value]))

def int64_list_feature(value):
  return tf.train.Feature(int64_list=tf.train.Int64List(value=value))

def create_tf_example(group, path):
    with tf.io.gfile.GFile(os.path.join(path, '{}'.format(group.filename)), 'rb') as fid:
        encoded_jpg = fid.read()
    encoded_jpg_io = io.BytesIO(encoded_jpg)
    image = Image.open(encoded_jpg_io)
    key = hashlib.sha256(encoded_jpg).hexdigest()
    width, height = image.size

    filename = group.filename.encode('utf8')
    xmins = []
    xmaxs = []
    ymins = []
    ymaxs = []
    classes_text = []
    classes = []

    for index, row in group.object.iterrows():
        xmins.append(row['xmin'] / width)
        xmaxs.append(row['xmax'] / width)
        ymins.append(row['ymin'] / height)
        ymaxs.append(row['ymax'] / height)
        classes_text.append(row['class'].encode('utf8'))
        classes.append(config.DETECTOR_CLASSES.index(row['class'])+1)

    tf_example = tf.train.Example(features=tf.train.Features(feature={
      'image/height': int64_feature(height),
      'image/width': int64_feature(width),
      'image/filename': bytes_feature(filename),
      'image/source_id': bytes_feature(filename),
      'image/key/sha256': bytes_feature(key.encode('utf8')),
      'image/encoded': bytes_feature(encoded_jpg),
      'image/format': bytes_feature('jpeg'.encode('utf8')),
      'image/object/bbox/xmin': float_list_feature(xmins),
      'image/object/bbox/xmax': float_list_feature(xmaxs),
      'image/object/bbox/ymin': float_list_feature(ymins),
      'image/object/bbox/ymax': float_list_feature(ymaxs),
      'image/object/class/text': bytes_list_feature(classes_text),
      'image/object/class/label': int64_list_feature(classes),
    }))

    return tf_example

def generate_tfrecord():
    writer = tf.io.TFRecordWriter(config.DETECTOR_DATASET_TFRECORD_PATH)
    path = os.path.join(config.DETECTOR_DATASET_PATH)
    examples = pd.read_csv(config.DETECTOR_DATASET_CSV_PATH)
    grouped = split(examples, 'filename')
    for group in grouped:
        tf_example = create_tf_example(group, path)
        writer.write(tf_example.SerializeToString())

    writer.close()
    output_path = os.path.join(os.getcwd(), config.DETECTOR_DATASET_TFRECORD_PATH)
    print('Successfully created the TFRecords: {}'.format(output_path))


if __name__ == '__main__':
    dataset_xml = xml_to_csv(config.DETECTOR_DATASET_PATH)
    dataset_xml.to_csv(config.DETECTOR_DATASET_CSV_PATH,index=None)
    generate_tfrecord()
```

**Comment by DeltaTesting** [comment]
After looking arround and some prints here and there, i've almost discarted that my tfrecords where wrong, all data that i pass to the tf.train.Example(..) is correct. 
The error that I still have looks like it happens when the loss function is called, so the output of the network is '-1' when it should be a number between 0 and 80 and the loss can't be computed.

The line where the error happens in the train.py when executed with `--mode eager_tf` is '127', where the loss_fn is called for each output.

Any ideas that why this happens?

**Comment by DeltaTesting** [comment]
After looking arround and some prints here and there, i've seen where's the error, i use the same value for 'image/object/class/text' and 'image/object/class/label', in my case the object ID, but the script exppect the text to be the name of the object and then use an ID calculated with the class names file.

I've modified the script (dataset.py) to use the label value instead and all works fine now.

` def parse_tfrecord(tfrecord, class_table):

      x = tf.io.parse_single_example(tfrecord, IMAGE_FEATURE_MAP)
      x_train = tf.image.decode_jpeg(x['image/encoded'], channels=3)
      x_train = tf.image.resize(x_train, (416, 416))
      # class_text = tf.sparse.to_dense(x['image/object/class/text'], default_value='')
      # labels = tf.cast(class_table.lookup(class_text), tf.float32)
      labels = tf.cast(tf.sparse.to_dense( x['image/object/class/label'], default_value=0), tf.float32)
      y_train = tf.stack([tf.sparse.to_dense(x['image/object/bbox/xmin']),
                        tf.sparse.to_dense(x['image/object/bbox/ymin']),
                        tf.sparse.to_dense(x['image/object/bbox/xmax']),
                        tf.sparse.to_dense(x['image/object/bbox/ymax']),
                        labels], axis=1)
    paddings = [[0, 100 - tf.shape(y_train)[0]], [0, 0]]
    y_train = tf.pad(y_train, paddings)
    return x_train, y_train`

**Comment by AnaRhisT94** [comment]
Thanks @DeltaTesting , I've changed this too now but it's still not training. I have `nan` values.
Would love to get some help with this.
My `.txt` files look like this:
```
0 0.75078125 0.26481481481481484 0.05364583333333333 0.024074074074074074
1 0.7997395833333333 0.21898148148148147 0.2921875 0.43796296296296294
```
Where `0` is `plates` class, and `1` is vehicles class.

My `create_tfrecord.py` file:

```
import os
import hashlib
import random
from pathlib import Path

import cv2 as cv
import tensorflow as tf
from absl import app


# train_txt = './train.txt'
# images_path = './JPEGImages'


def bytes_feature(value):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))


def bytes_list_feature(value):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=value))


def float_list_feature(value):
    return tf.train.Feature(float_list=tf.train.FloatList(value=value))


def int64_feature(value):
    return tf.train.Feature(int64_list=tf.train.Int64List(value=[value]))


def int64_list_feature(value):
    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))


def get_example(line, image_path):
    class_text = []
    labels = []
    xmin = []
    ymin = []
    xmax = []
    ymax = []
    truncated = []
    poses = []
    difficult_obj = []

    # read the image
    with tf.io.gfile.GFile(image_path, 'rb') as fib:
        image_encoded = fib.read()
    h, w, _ = cv.imread(image_path).shape
    key = hashlib.sha256(image_encoded).hexdigest()

    # read the parameters
    item = line.split(' ')
    if item[0] == "0":
        # class 0 is plates
        class_text.append("plates".encode('utf8'))
        labels.append(0)
    elif item[0] == "1":
        # class 1 is vehicles
        class_text.append("vehicles".encode('utf8'))
        labels.append(1)
    else:
        raise ValueError("Error! I expect to get `0` for plates and `1` "
                         "for vehicles. I got {} instead.".format(item[0]))

    xmin.append(float(item[1]))
    ymin.append(float(item[2]))
    xmax.append(float(item[3]))
    ymax.append(float(item[4]))

    truncated.append(0)
    difficult_obj.append(0)
    poses.append("Unspecified".encode('utf8'))

    example = tf.train.Example(features=tf.train.Features(feature={
        'image/height': int64_feature(h),
        'image/width': int64_feature(w),
        'image/filename': bytes_feature(image_path.encode('utf8')),
        'image/source_id': bytes_feature(image_path.encode('utf8')),
        'image/key/sha256': bytes_feature(key.encode('utf8')),
        'image/encoded': bytes_feature(image_encoded),
        'image/format': bytes_feature('png'.encode('utf8')),
        'image/object/bbox/xmin': float_list_feature(xmin),
        'image/object/bbox/xmax': float_list_feature(xmax),
        'image/object/bbox/ymin': float_list_feature(ymin),
        'image/object/bbox/ymax': float_list_feature(ymax),
        'image/object/class/text': bytes_list_feature(class_text),
        'image/object/class/label': int64_list_feature(labels),
    }))

    return example


def create_tfrecord():
    # Create a TFRecordWriter, we will save the outputs in the following file
    train_writer = tf.io.TFRecordWriter('./tf_record_big/train.tfrecord')
    val_writer = tf.io.TFRecordWriter('./tf_record_big/val.tfrecord')

    # Counter to count the number of training and validation samples
    train_ctr = val_ctr = 0

    image_format_train = img_format_val = ".png"

    # list all the .txt files in the current folder (was previously ./mini_training_set)
    training_txt_files = [f for f in os.listdir('/home/me/Desktop/training_data/mini_training_set') if f.endswith('.txt')]
    txt_files_length = len(training_txt_files)
    # loop over all the training .txt files
    for i, txt_file in enumerate(training_txt_files):
        print("Working on file: {} / {}".format(i, txt_files_length))
        # get the associated image's file name and path
        image_name = txt_file.split(".")[0] + image_format_train
        image_path = os.path.join('/home/me/Desktop/training_data/mini_training_set', image_name)
        txt_path = os.path.join('/home/me/Desktop/training_data/mini_training_set', txt_file)
        if len(image_name) == 0 or len(image_path) == 0 or len(txt_path) == 0:
            continue
        if os.path.isfile(image_path) is False:
            print("Skipped that non-existant image.")
            continue
        with open(txt_path, 'r') as train_file:
            lines = train_file.read().split('\n')
            # random.shuffle(lines)
            # loop over all the training file lines
            for line in lines:
                if len(line) > 0:
                    example = get_example(line, image_path)
                    train_ctr += 1
                    train_writer.write(example.SerializeToString())
            train_file.close()
    print("Finishing reading the training .txt files.")
    print("Preparing to read now the validation .txt files.")

    # Prepare for creating TFRecord for validation set
    validation_txt_files = [f for f in os.listdir('/home/me/Desktop/training_data/mini_validation_set_same_as_training') if f.endswith('.txt')]
    val_files_length = len(validation_txt_files)
    # loop over all the training .txt files
    for i, txt_file in enumerate(validation_txt_files):
        print("Working on file: {} / {}".format(i, val_files_length))
        # get the associated image's file name and path
        image_name = txt_file.split(".")[0] + img_format_val
        if len(image_name) == 0 or len(image_path) == 0 or len(txt_path) == 0:
            continue
        image_path = os.path.join('/home/me/Desktop/training_data/mini_validation_set_same_as_training', image_name)
        txt_path = os.path.join('/home/me/Desktop/training_data/mini_validation_set_same_as_training', txt_file)
        if os.path.isfile(image_path) is False:
            print("Skipped that non-existant image.")
            continue
        with open(txt_path, 'r') as val_file:
            lines = val_file.read().split('\n')
            # random.shuffle(lines)
            # loop over all the training file lines
            for line in lines:
                if len(line) > 0:
                    example = get_example(line, image_path)
                    val_ctr += 1
                    val_writer.write(example.SerializeToString())
            val_file.close()

    print("Length of validation set: {}".format(val_ctr))
    print("Length of training set: {}".format(train_ctr))
    print("Finished reading all text files!")
    return True


if __name__ == '__main__':
    try:
        app.run(create_tfrecord())
    except SystemExit:
        pass
```

**Comment by AnaRhisT94** [comment]
@DeltaTesting I've created a thread for this if you can take a look it'd be highly appreciated:
https://github.com/zzh8829/yolov3-tf2/issues/75
